swupd/manifest.go

```diff
@@ -319,3 +342,215 @@ func (m *Manifest) WriteManifestFile(path string) error {
 	// a nil error may be replaced by an f.Close() error in the deferred function
 	return err
 }
+
+func (m *Manifest) sortFilesName() {
+	// files
+	sort.Slice(m.Files, func(i, j int) bool {
+		return m.Files[i].Name < m.Files[j].Name
+	})
+	// and deleted files
+	sort.Slice(m.DeletedFiles, func(i, j int) bool {
+		return m.DeletedFiles[i].Name < m.DeletedFiles[j].Name
+	})
+}
+
+func (m *Manifest) sortFilesVersionName() {
+	// files
+	sort.Slice(m.Files, func(i, j int) bool {
+		if m.Files[i].Version < m.Files[j].Version {
+			return true
+		}
+		if m.Files[i].Version > m.Files[j].Version {
+			return false
+		}
+		return m.Files[i].Name < m.Files[j].Name
+	})
+	// and deleted files
+	sort.Slice(m.DeletedFiles, func(i, j int) bool {
+		if m.DeletedFiles[i].Version < m.DeletedFiles[j].Version {
+			return true
+		}
+		if m.DeletedFiles[i].Version > m.DeletedFiles[j].Version {
+			return false
+		}
+		return m.DeletedFiles[i].Name < m.DeletedFiles[j].Name
+	})
+}
+
+func (m *Manifest) addDeleted(oldManifest *Manifest) {
+	for _, df := range oldManifest.DeletedFiles {
+		if df.findFileNameInSlice(m.Files) == nil {
+			m.Files = append(m.Files, df)
+			m.DeletedFiles = append(m.DeletedFiles, df)
+		}
+	}
+}
+
+// linkPeersAndChange
+// At this point Manifest m should only have the files that were present
+// in the chroot for that manifest. Link delta peers with the oldManifest
+// if the file in the oldManifest is not deleted or ghosted
+func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) int {
+	changed := 0
+	for _, of := range oldManifest.Files {
+		if match := of.findFileNameInSlice(m.Files); match != nil {
+			if of.Status == statusDeleted || of.Status == statusGhosted {
+				continue
+			}
+
+			match.DeltaPeer = of
+			of.DeltaPeer = match
+			if !sameFile(match, of) {
+				match.Version = m.Header.Version
+				changed++
+			} else {
+				match.Version = of.Version
+			}
+		}
+	}
+
+	return changed
+}
+
+func (m *Manifest) filesAdded(oldManifest *Manifest) int {
+	added := 0
+	for _, af := range m.Files {
+		if af.findFileNameInSlice(oldManifest.Files) == nil {
+			af.Version = m.Header.Version
+			added++
+		}
+	}
+
+	return added
+}
+
+func (m *Manifest) newDeleted(oldManifest *Manifest) int {
```

Comment:
Optional: filesAdded, newDeleted and linkPeersAndChanged could be calculated in the same pass using two sorted list of files, like the original swupd-server did (and this code does elsewhere).